doc/generated/pmemobj_tx_add_range.3
.\" Automatically generated by Pandoc 2.0.6
.\"
.TH "PMEMOBJ_TX_ADD_RANGE" "3" "2020-01-16" "PMDK - pmemobj API version 2.3" "PMDK Programmer's Manual"
.hy
.\" Copyright 2014-2020, Intel Corporation
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in
.\"       the documentation and/or other materials provided with the
.\"       distribution.
.\"
.\"     * Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SH NAME
.PP
\f[B]pmemobj_tx_add_range\f[](), \f[B]pmemobj_tx_add_range_direct\f[](),
\f[B]pmemobj_tx_xadd_range\f[](),
\f[B]pmemobj_tx_xadd_range_direct\f[]()
.PP
\f[B]TX_ADD\f[](), \f[B]TX_ADD_FIELD\f[](), \f[B]TX_ADD_DIRECT\f[](),
\f[B]TX_ADD_FIELD_DIRECT\f[](),
.PP
\f[B]TX_XADD\f[](), \f[B]TX_XADD_FIELD\f[](), \f[B]TX_XADD_DIRECT\f[](),
\f[B]TX_XADD_FIELD_DIRECT\f[](),
.PP
\f[B]TX_SET\f[](), \f[B]TX_SET_DIRECT\f[](), \f[B]TX_MEMCPY\f[](),
\f[B]TX_MEMSET\f[]() \- transactional object manipulation
.SH SYNOPSIS
.IP
.nf
\f[C]
#include\ <libpmemobj.h>

int\ pmemobj_tx_add_range(PMEMoid\ oid,\ uint64_t\ off,\ size_t\ size);
int\ pmemobj_tx_add_range_direct(const\ void\ *ptr,\ size_t\ size);
int\ pmemobj_tx_xadd_range(PMEMoid\ oid,\ uint64_t\ off,\ size_t\ size,\ uint64_t\ flags);
int\ pmemobj_tx_xadd_range_direct(const\ void\ *ptr,\ size_t\ size,\ uint64_t\ flags);

TX_ADD(TOID\ o)
TX_ADD_FIELD(TOID\ o,\ FIELD)
TX_ADD_DIRECT(TYPE\ *p)
TX_ADD_FIELD_DIRECT(TYPE\ *p,\ FIELD)

TX_XADD(TOID\ o,\ uint64_t\ flags)
TX_XADD_FIELD(TOID\ o,\ FIELD,\ uint64_t\ flags)
TX_XADD_DIRECT(TYPE\ *p,\ uint64_t\ flags)
TX_XADD_FIELD_DIRECT(TYPE\ *p,\ FIELD,\ uint64_t\ flags)

TX_SET(TOID\ o,\ FIELD,\ VALUE)
TX_SET_DIRECT(TYPE\ *p,\ FIELD,\ VALUE)
TX_MEMCPY(void\ *dest,\ const\ void\ *src,\ size_t\ num)
TX_MEMSET(void\ *dest,\ int\ c,\ size_t\ num)
\f[]
.fi
.SH DESCRIPTION
.PP
\f[B]pmemobj_tx_add_range\f[]() takes a \[lq]snapshot\[rq] of the memory
block of given \f[I]size\f[], located at given offset \f[I]off\f[] in
the object specified by \f[I]oid\f[], and saves it to the undo log.
The application is then free to directly modify the object in that
memory range.
In case of a failure or abort, all the changes within this range will be
rolled back.
The supplied block of memory has to be within the pool registered in the
transaction.
This function must be called during \f[B]TX_STAGE_WORK\f[].
.PP
The \f[B]pmemobj_tx_xadd_range\f[]() function behaves exactly the same
as \f[B]pmemobj_tx_add_range\f[]() when \f[I]flags\f[] equals zero.
\f[I]flags\f[] is a bitmask of the following values:
.IP \[bu] 2
\f[B]POBJ_XADD_NO_FLUSH\f[] \- skip flush on commit (when application
deals with flushing or uses pmemobj_memcpy_persist)
.IP \[bu] 2
\f[B]POBJ_XADD_NO_SNAPSHOT\f[] \- added range will not be
\[lq]snapshotted\[rq], i.e.\ any changes made within it during the
transaction will not be rolled backed after abort
.IP \[bu] 2
\f[B]POBJ_XADD_ASSUME_INITIALIZED\f[] \- added range is assumed to be
initialized.
If this flag is not specified, passing uninitialized memory will result
in an error when run under Valgrind memcheck.
.PP
\f[B]pmemobj_tx_add_range_direct\f[]() behaves the same as
\f[B]pmemobj_tx_add_range\f[]() with the exception that it operates on
virtual memory addresses and not persistent memory objects.
It takes a \[lq]snapshot\[rq] of a persistent memory block of given
\f[I]size\f[], located at the given address \f[I]ptr\f[] in the virtual
memory space and saves it to the undo log.
The application is then free to directly modify the object in that
memory range.
In case of a failure or abort, all the changes within this range will be
rolled back.
The supplied block of memory has to be within the pool registered in the
transaction.
This function must be called during \f[B]TX_STAGE_WORK\f[].
.PP
The \f[B]pmemobj_tx_xadd_range_direct\f[]() function behaves exactly the
same as \f[B]pmemobj_tx_add_range_direct\f[]() when \f[I]flags\f[]
equals zero.
\f[I]flags\f[] is a bitmask of the following values:
.IP \[bu] 2
\f[B]POBJ_XADD_NO_FLUSH\f[] \- skip flush on commit (when application
deals with flushing or uses pmemobj_memcpy_persist)
.IP \[bu] 2
\f[B]POBJ_XADD_NO_SNAPSHOT\f[] \- added range will not be
\[lq]snapshotted\[rq], i.e.\ any changes made within it during the
transaction will not be rolled backed after abort
.IP \[bu] 2
\f[B]POBJ_XADD_ASSUME_INITIALIZED\f[] \- added range is assumed to be
initialized.
If this flag is not specified, passing uninitialized memory will result
in an error when run under Valgrind memcheck.
.PP
Similarly to the macros controlling the transaction flow,
\f[B]libpmemobj\f[] defines a set of macros that simplify the
transactional operations on persistent objects.
Note that those macros operate on typed object handles, thus eliminating
the need to specify the size of the object, or the size and offset of
the field in the user\-defined structure that is to be modified.
.PP
The \f[B]TX_ADD_FIELD\f[]() macro saves the current value of given
\f[I]FIELD\f[] of the object referenced by a handle \f[I]o\f[] in the
undo log.
The application is then free to directly modify the specified
\f[I]FIELD\f[].
In case of a failure or abort, the saved value will be restored.
.PP
The \f[B]TX_XADD_FIELD\f[]() macro works exactly like
\f[B]TX_ADD_FIELD\f[] when \f[I]flags\f[] equals 0.
The \f[I]flags\f[] argument is a bitmask of values described in
\f[B]pmemobj_tx_xadd_range\f[], above.
.PP
The \f[B]TX_ADD\f[]() macro takes a \[lq]snapshot\[rq] of the entire
object referenced by object handle \f[I]o\f[] and saves it in the undo
log.
The object size is determined from its \f[I]TYPE\f[].
The application is then free to directly modify the object.
In case of a failure or abort, all the changes within the object will be
rolled back.
.PP
The \f[B]TX_XADD\f[]() macro works exactly like \f[B]TX_ADD\f[] when
\f[I]flags\f[] equals 0.
The \f[I]flags\f[] argument is a bitmask of values as described in
\f[B]pmemobj_tx_xadd_range\f[], above.
.PP
The \f[B]TX_ADD_FIELD_DIRECT\f[]() macro saves the current value of the
given \f[I]FIELD\f[] of the object referenced by (direct) pointer
\f[I]p\f[] in the undo log.
The application is then free to directly modify the specified
\f[I]FIELD\f[].
In case of a failure or abort, the saved value will be restored.
.PP
The \f[B]TX_XADD_FIELD_DIRECT\f[]() macro works exactly like
\f[B]TX_ADD_FIELD_DIRECT\f[] when \f[I]flags\f[] equals 0.
The \f[I]flags\f[] argument is a bitmask of values as described in
\f[B]pmemobj_tx_xadd_range_direct\f[], above.
.PP
The \f[B]TX_ADD_DIRECT\f[]() macro takes a \[lq]snapshot\[rq] of the
entire object referenced by (direct) pointer \f[I]p\f[] and saves it in
the undo log.
The object size is determined from its \f[I]TYPE\f[].
The application is then free to directly modify the object.
In case of a failure or abort, all the changes within the object will be
rolled back.
.PP
The \f[B]TX_XADD_DIRECT\f[]() macro works exactly like
\f[B]TX_ADD_DIRECT\f[] when \f[I]flags\f[] equals 0.
The \f[I]flags\f[] argument is a bitmask of values as described in
\f[B]pmemobj_tx_xadd_range_direct\f[], above.
.PP
The \f[B]TX_SET\f[]() macro saves the current value of the given
\f[I]FIELD\f[] of the object referenced by handle \f[I]o\f[] in the undo
log, and then sets its new \f[I]VALUE\f[].
In case of a failure or abort, the saved value will be restored.
.PP
The \f[B]TX_SET_DIRECT\f[]() macro saves in the undo log the current
value of given \f[I]FIELD\f[] of the object referenced by (direct)
pointer \f[I]p\f[], and then set its new \f[I]VALUE\f[].
In case of a failure or abort, the saved value will be restored.
.PP
The \f[B]TX_MEMCPY\f[]() macro saves in the undo log the current content
of \f[I]dest\f[] buffer and then overwrites the first \f[I]num\f[] bytes
of its memory area with the data copied from the buffer pointed by
\f[I]src\f[].
In case of a failure or abort, the saved value will be restored.
.PP
The \f[B]TX_MEMSET\f[]() macro saves the current content of the
\f[I]dest\f[] buffer in the undo log and then fills the first
\f[I]num\f[] bytes of its memory area with the constant byte \f[I]c\f[].
In case of a failure or abort, the saved value will be restored.
.SH RETURN VALUE
.PP
On success, \f[B]pmemobj_tx_add_range\f[](),
\f[B]pmemobj_tx_xadd_range\f[](), \f[B]pmemobj_tx_add_range_direct\f[]()
and \f[B]pmemobj_tx_xadd_range_direct\f[]() return 0.
Otherwise, the stage is changed to \f[B]TX_STAGE_ONABORT\f[] and an
error number is returned.
.SH SEE ALSO
.PP
\f[B]pmemobj_tx_alloc\f[](3), \f[B]pmemobj_tx_begin\f[](3),
\f[B]libpmemobj\f[](7) and \f[B]<http://pmem.io>\f[]


ChangeLog
Fri Jan 17 2020 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.8-rc1

	This release:
	- Introduces new API in libpmemobj for user-assisted defragmentation
	  (see pmemobj_defrag man page for details).
	- Introduces experimental support for PPC64LE architecture.
	- Introduces new flag in libpmemobj for opting-out of transaction aborts
	  on failure (POBJ_TX_NO_ABORT), along with new variants of existing APIs
	  that didn't accept flags (pmemobj_tx_xfree, pmemobj_tx_xpublish,
	  pmemobj_tx_xlock, pmemobj_tx_xlog_append_buffer, pmemobj_tx_xstrdup,
	  pmemobj_tx_xwcsdup).
	- Moves out libvmem and libvmmalloc to the new repository
	  (https://github.com/pmem/vmem).

	Other changes:
	- obj: introduce new statistics useful for defrag
	- obj: introduce transient/persistent statistics enabling
	- obj: introduce pmemobj_tx_(set/get)_user_data funcs
	- obj: introduce pmemobj_(set/get)_user_data funcs
	- obj: disable workaround for offsetof() since VS 15.5 in public header
	- common: drop support for libndctl < 63 on Linux
	- pool: rename -N --no-exec to -d --dry-run

	Notable bug fixes:
	- obj: fix potential NULL-dereference in ulog_store
	- obj: fix unintended recursive locking during palloc
	- obj: fix lock release order in palloc publish
	- obj: fix transient redo log of more than 64 entries
	- obj: fix capacity ulog calculation
	- obj: fix check of unaligned capacity size
	- rpmem: add a missing case for GPSPM + FLUSH_STRICT
	- pmem: fix pmemcheck support on platforms with eADR
	- pool: fix possible memory leak
	- rpmem: fix possible memory leak in rpmemd_config_read
	- rpmem: fix possible memory leak in rpmemd_log_init
	- rpmem: fix possible use-after-free

Mon Sep 30 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.7

	This release:
	- Introduces new APIs in libpmemobj for managing space used by transactions.
	  (see pmemobj_tx_log_append_buffer man page for details)
	- Introduces new APIs in librpmem, splitting rpmem_persist into rpmem_flush
	  and rpmem_drain, allowing applications to use the flush + drain model
	  already known from libpmem. (libpmemobj does not use this feature yet)
	- Optimizes large libpmemobj transactions by significantly reducing
	  the amount of memory modified at the commit phase.
	- Optimizes tracking of libpmemobj reservations.
	- Adds new flags for libpmemobj's pmemobj_tx_xadd_range[_direct] API:
	  POBJ_XADD_NO_SNAPSHOT and POBJ_XADD_ASSUME_INITIALIZED, allowing
	  applications to optimize how memory is tracked by the library.

	To support some of the above changes the libpmemobj on-media layout had
	to be changed, which means that old pools have to be converted using
	pmdk-convert >= 1.7.

	Other changes:
	- obj: fix merging of ranges when NOFLUSH flag is used (#1100)
	- rpmem: fix closing of ssh connection (#995, #1060)
	- obj: abort transaction on pmemobj_tx_publish failure

	Internal changes:
	- test: fault injection tests for pmemblk, pmemlog, and pmemobj
	- test: improved Python testing framework
	- test: support real pmem in bad blocks tests
	- common: allow not building examples and benchmarks

Tue Aug 27 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.6.1

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains,
	incorporates build fixes for FreeBSD and fixes a number of
	smaller bugs.

	Detailed list of bug fixes:
	- common: (win) fix possible pool file coruption (#972, #715, #603)
	- common: implement correct / robust device_dax_alignment (#1071)
	- obj: fix recycler not locating unused chunks
	- doc: update pmemobj_tx_lock documentation wrt behavior on fail
	- common: fix persistent domain detection (#1093)
	- common: vecq: fix a pointer-to-struct aliasing violation
	  (crash on arm64)
	- common: fix minor issues related to ndctl linking
	- obj: drop recursion from pmemobj_next
	- common: fix bug in badblock file error handling
	- obj: fix handling of malloc failures
	- common: fix handling of malloc failures (ctl)
	- jemalloc: fix build with gcc 9
	- obj: don't overwrite errno when palloc_heap_check_remote fails
	- doc: fix pmreorder emit log macro
	- rpmem: change order of rpmem init (#1043)
	- common: Fix build failure due to unused macro PAGE_SIZE
	- common: support older versions of pkg-config
	- tools: link with release variant of pmemcommon
	- common: add PMDK prefix to local copy of queue.h (#990)
	- rpmem: switch to using an open coded basename (FreeBSD)
	- common: posix_fallocate: guard against integer underflow in
	  check (FreeBSD)
	- test: support Valgrind 3.15
	- test: skip if fi_info is missing
	- test: (win) fix sparsefile error handling
	- test: fix libpmempool_feature tests that match logs
	- test: remove vmem_delete test (#1074)
	- test: adjust matchfiles in vmem_valgrind_region test (#1087)
	- test: remove old log files for windows (#1013)
	- test: remove invalid expect_normal_exit (#1092)
	- test: suppress ld leak (#1098)
	- test: Expose necessary symbols in libvmmalloc_dummy_funcs
	  (FreeBSD)
	- test: fix tests failing because `tput` fails (FreeBSD)
	- test: avoid obj_critnib_mt taking very long on many-core machines
	- test: deal with libndctl's path without build system
	- test: overwrite old log in pmempool_create/TEST14.PS1
	- test: fix match files in tests which use dax devices
	- test: fix match file in rpmem_addr_ext test
	- test: fix pmempool_check test

Wed Aug 28 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.5.2

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains

src/benchmarks/pmemobj_tx_add_range.cpp
// SPDX-License-Identifier: BSD-3-Clause
/* Copyright 2019, Intel Corporation */

/*
 * pmemobj_tx_add_range.cpp -- pmemobj_tx_add_range benchmarks definition
 */

#include <cassert>
#include <cerrno>
#include <cstddef>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fcntl.h>
#include <sys/file.h>
#include <sys/mman.h>
#include <unistd.h>

#include "benchmark.hpp"
#include "file.h"
#include "libpmemobj.h"

#define LAYOUT_NAME "tx_add_range_benchmark"

/*
 * POOL_SIZE_COEFFICIENT -- pool has to hold every allocated object with
 * its snapshot (1 + 1), plus 0.5 because of fragmentation
 */
#define POOL_SIZE_COEFFICIENT (1 + 1 + 0.5)

/*
 * MAX_ALLOC_SIZE -- maximum size of one allocation (128 MiB)
 */
#define MAX_ALLOC_SIZE (1024 * 1024 * 128)

/*
 * ranged_obj -- ranged object
 */
struct ranged_obj {
	void *ptr;   /* address of allocated object */
	size_t size; /* size of allocated object */
};

/*
 * obj_bench_args -- benchmark specific command line options
 */
struct obj_bench_args {
	uint64_t nranges;  /* number of allocated objects */
	bool shuffle_objs; /* shuffles the array of allocated objects */
};

/*
 * obj_bench -- benchmark context
 */
struct obj_bench {
	PMEMobjpool *pop;	  /* persistent pool handle */
	struct ranged_obj *ranges; /* array of ranges */
	size_t obj_size;	   /* size of a single range */
	uint64_t nranges;	  /* number of ranges */
	uint64_t nallocs;	  /* number of allocations */
	bool shuffle_objs;	 /* shuffles array of ranges */
	unsigned seed;		   /* PRNG seed */
};

/*
 * shuffle_ranges -- randomly shuffles elements in an array
 * to avoid sequential pattern in the transaction loop
 */
static void
shuffle_ranges(struct ranged_obj *ranged, uint64_t nranges, unsigned *seed)
{
	struct ranged_obj tmp;
	uint64_t dest;

	for (uint64_t n = 0; n < nranges; ++n) {
		dest = RRAND_R(seed, nranges - 1, 0);
		tmp = ranged[n];
		ranged[n] = ranged[dest];
		ranged[dest] = tmp;
	}
}

/*
 * init_ranges -- allocate persistent objects and carve ranges from them
 */
static int
init_ranges(struct obj_bench *ob)
{
	assert(ob->nranges != 0);
	ob->ranges = (struct ranged_obj *)malloc((ob->nranges) *
						 sizeof(struct ranged_obj));
	if (!ob->ranges) {
		perror("malloc");
		return -1;
	}

	size_t nranges_per_object = MAX_ALLOC_SIZE / ob->obj_size;

	for (size_t i = 0, n = 0; n < ob->nranges && i < ob->nallocs; i++) {
		PMEMoid oid;
		if (pmemobj_alloc(ob->pop, &oid, MAX_ALLOC_SIZE, 0, nullptr,
				  nullptr)) {
			perror("pmemobj_alloc");
			goto err;
		}

		for (size_t j = 0; j < nranges_per_object; j++) {
			void *ptr = (char *)pmemobj_direct(oid) +
				(j * ob->obj_size);
			struct ranged_obj range = {ptr, ob->obj_size};
			ob->ranges[n++] = range;
			if (n == ob->nranges)
				break;
		}
	}

	if (ob->shuffle_objs == true)
		shuffle_ranges(ob->ranges, ob->nranges, &ob->seed);

	return 0;

err:
	free(ob->ranges);
	return -1;
}

/*
 * tx_add_range_init -- initialization function
 */
static int
tx_add_range_init(struct benchmark *bench, struct benchmark_args *args)
{
	assert(bench != nullptr);
	assert(args != nullptr);
	assert(args->opts != nullptr);

	struct obj_bench_args *bargs = (struct obj_bench_args *)args->opts;

	enum file_type type = util_file_get_type(args->fname);
	if (type == OTHER_ERROR) {
		fprintf(stderr, "could not check type of file %s\n",
			args->fname);
		return -1;
	}

	auto *ob = (struct obj_bench *)malloc(sizeof(struct obj_bench));
	if (ob == nullptr) {
		perror("malloc");
		return -1;
	}

	/* let's calculate number of allocations */
	ob->nallocs = (args->dsize * bargs->nranges / MAX_ALLOC_SIZE) + 1;

	size_t pool_size;

	if (args->is_poolset || type == TYPE_DEVDAX)
		pool_size = 0;
	else {
		pool_size =
			ob->nallocs * MAX_ALLOC_SIZE * POOL_SIZE_COEFFICIENT;
	}

	/* create pmemobj pool */
	ob->pop = pmemobj_create(args->fname, LAYOUT_NAME, pool_size,
				 args->fmode);
	if (ob->pop == nullptr) {
		fprintf(stderr, "%s\n", pmemobj_errormsg());
		goto err;
	}

	ob->nranges = bargs->nranges;
	ob->obj_size = args->dsize;
	ob->shuffle_objs = bargs->shuffle_objs;
	ob->seed = args->seed;

	if (init_ranges(ob))
		goto err_pop_close;

	pmembench_set_priv(bench, ob);

	return 0;

err_pop_close:
	pmemobj_close(ob->pop);
err:
	free(ob);
	return -1;
}

/*
 * tx_add_range_op -- actual benchmark operation
 */
static int
tx_add_range_op(struct benchmark *bench, struct operation_info *info)
{
	auto *ob = (struct obj_bench *)pmembench_get_priv(bench);
	int ret = 0;

	TX_BEGIN(ob->pop)
	{
		for (size_t i = 0; i < ob->nranges; i++) {
			struct ranged_obj *r = &ob->ranges[i];
			pmemobj_tx_add_range_direct(r->ptr, r->size);
		}
	}
	TX_ONABORT
	{
		fprintf(stderr, "transaction failed\n");
		ret = -1;
	}
	TX_END

	return ret;
}

/*
 * tx_add_range_exit -- benchmark cleanup function
 */
static int
tx_add_range_exit(struct benchmark *bench, struct benchmark_args *args)
{
	auto *ob = (struct obj_bench *)pmembench_get_priv(bench);

	pmemobj_close(ob->pop);
	free(ob->ranges);
	free(ob);

	return 0;
}

static struct benchmark_clo tx_add_range_clo[2];

/* Stores information about benchmark. */
static struct benchmark_info tx_add_range_info;
CONSTRUCTOR(tx_add_range_constructor)
void
tx_add_range_constructor(void)
{
	tx_add_range_clo[0].opt_short = 0;
	tx_add_range_clo[0].opt_long = "num-of-ranges";
	tx_add_range_clo[0].descr = "Number of ranges";
	tx_add_range_clo[0].def = "1000";
	tx_add_range_clo[0].off =
		clo_field_offset(struct obj_bench_args, nranges);
	tx_add_range_clo[0].type = CLO_TYPE_UINT;
	tx_add_range_clo[0].type_uint.size =
		clo_field_size(struct obj_bench_args, nranges);
	tx_add_range_clo[0].type_uint.base = CLO_INT_BASE_DEC;
	tx_add_range_clo[0].type_uint.min = 1;
	tx_add_range_clo[0].type_uint.max = ULONG_MAX;

	tx_add_range_clo[1].opt_short = 's';
	tx_add_range_clo[1].opt_long = "shuffle";
	tx_add_range_clo[1].descr =
		"Use shuffle objects - "
		"randomly shuffles array of allocated objects";
	tx_add_range_clo[1].def = "false";
	tx_add_range_clo[1].off =
		clo_field_offset(struct obj_bench_args, shuffle_objs);
	tx_add_range_clo[1].type = CLO_TYPE_FLAG;

	tx_add_range_info.name = "pmemobj_tx_add_range";
	tx_add_range_info.brief = "Benchmark for pmemobj_tx_add_range() "
				  "operation";
	tx_add_range_info.init = tx_add_range_init;
	tx_add_range_info.exit = tx_add_range_exit;
	tx_add_range_info.multithread = true;
	tx_add_range_info.multiops = true;
	tx_add_range_info.operation = tx_add_range_op;
	tx_add_range_info.measure_time = true;
	tx_add_range_info.clos = tx_add_range_clo;
	tx_add_range_info.nclos = ARRAY_SIZE(tx_add_range_clo);
	tx_add_range_info.opts_size = sizeof(struct obj_bench_args);
	tx_add_range_info.rm_file = true;
	tx_add_range_info.allow_poolset = true;
	REGISTER_BENCHMARK(tx_add_range_info);
};
